### ISSUE TITLE
Simple model cannot converge with fp16 precision using torch.cuda.amp

### ISSUE THREAD

**Issue by back2yes** [pytorch/pytorch#46238]
## To Reproduce

```
import torch
from torch import nn
import torch.nn.functional as F


class SimpleNet(nn.Module):
    def __init__(self):
        super(SimpleNet, self).__init__()
        self.conv1 = nn.Conv2d(3, 16, 4, 2, 1)
        self.conv2 = nn.Conv2d(16, 16, 4, 2, 1)
        self.conv3 = nn.Conv2d(16, 16, 4, 2, 1)
        self.conv4 = nn.Conv2d(16, 16, 4, 2, 1)
        for m in self.modules():
            if isinstance(m, nn.Conv2d):
                # nn.init.zeros_(m.bias)
                nn.init.zeros_(m.bias)
                nn.init.xavier_normal_(m.weight, 1.0)

    def forward(self, x):
        x = self.conv1(x)
        x = F.relu(x)
        x = self.conv2(x)
        x = F.relu(x)
        x = self.conv3(x)
        x = F.relu(x)
        x = self.conv4(x)
        # x = F.relu(x)
        return x


batch_size = 32  # works for any batch_size <=63
# batch_size = 128  # FAILED to converge, any batch_size >= 64
base_lr = 5e-2
# lr = base_lr / 32 * batch_size  # No matter I rescale the lr w.r.t batch_size or not, model cannot converge with large batch.
lr = base_lr

scaler = torch.cuda.amp.GradScaler()
model = SimpleNet().cuda()
optimizer = torch.optim.SGD(model.parameters(), lr=lr, momentum=0.9)
for epoch in range(100):
    # X_batch = torch.zeros(10, 3, 128, 128).cuda()
    X_batch = torch.zeros(batch_size, 3, 128, 128).cuda()
    optimizer.zero_grad()

    # NEW
    with torch.cuda.amp.autocast():
        y_pred = model(X_batch).squeeze()
        loss = abs(y_pred - 1).mean()

    # NEW
    scaler.scale(loss).backward()
    lv = loss.detach().cpu().numpy()
    scaler.step(optimizer)
    scaler.update()

    print(f"epoch: {epoch}: \t{loss.item()}")

```

## Expected behavior

The model should converge with both batch_sizes of 32 and 128. However, with batch_size==32, the loss looks like:
epoch: 0: 	1.0
epoch: 1: 	1.0
epoch: 2: 	0.9970703125
epoch: 3: 	0.99072265625
epoch: 4: 	0.982421875
epoch: 5: 	0.9716796875
epoch: 6: 	0.958984375
epoch: 7: 	0.9443359375
epoch: 8: 	0.927734375
epoch: 9: 	0.91015625
epoch: 10: 	0.89111328125
epoch: 11: 	0.87060546875
epoch: 12: 	0.84912109375
epoch: 13: 	0.82666015625
epoch: 14: 	0.8037109375
epoch: 15: 	0.779296875
epoch: 16: 	0.7548828125
epoch: 17: 	0.7294921875
epoch: 18: 	0.703125
epoch: 19: 	0.6767578125
epoch: 20: 	0.6494140625
... ... 

But with larger batch_size (e.g. 128), the loss looks like:
epoch: 0: 	1.0
epoch: 1: 	1.0
epoch: 2: 	1.0
epoch: 3: 	1.0
epoch: 4: 	1.0
epoch: 5: 	1.0
epoch: 6: 	1.0
epoch: 7: 	1.0
epoch: 8: 	1.0
epoch: 9: 	1.0
epoch: 10: 	1.0
... ...

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

 - PyTorch Version (e.g., 1.0): 1.6
 - OS (e.g., Linux): Ubuntu 18.04
 - How you installed PyTorch (`conda`, `pip`, source):  `pip`
 - Build command you used (if compiling from source):
 - Python version: 3.7
 - CUDA/cuDNN version: 10.2
 - GPU models and configuration: NVIDIA 2080Ti
 - Any other relevant information:

## Additional context

I tried the code on 2 separate computers, the results are similar.

cc @mcarilli @ezyang

**Comment by bdhirsh** [comment]
I was able to repro this on 1.6.0, but not on tip-of-master. Seems like it's no longer an issue?

**Comment by back2yes** [comment]
@bdhirsh THX, I will try it.

**Comment by back2yes** [comment]
I tried building from source from `1.8.0a0+d5ca53c`, and it worked well with a batch_size as large as 1024. Great~
